Email notifications on incorrect app deployment.

The-Huginn opened the pull request:
Resolves #55

xstefank commented on src/main/java/io/xstefank/wildlfy/bot/LifecycleProcessor.java:
```diff
@@ -29,19 +31,36 @@ public class LifecycleProcessor {
+                        if (emailAddresses != null && !emailAddresses.isEmpty()) {LOG.infof("Sending email to the following people [%s].", String.join(", ", emailAddresses));
+                            mailer.send(
+                                    new Mail()
+                                            .setSubject("[Important] Unsuccessful installation of Wildfly Bot Application")
```
Drop important at the beggining

The-Huginn replied:
Will do

xstefank reviewed (commented):
looks good

The-Huginn commented on README.md:
```diff
@@ -78,6 +93,9 @@ wildfly:
      message: "Too many commits in PR!"
+ emails:
+   - xstefank@redhat.com
+   - petrberan@redhat.com
```
I screwed the emails as I only copied the names from `notify` field. I will change them it to your thing anyway.

The-Huginn commented:
I have committed latest changes into my branch but are not reflected in the PR. I suppose due to no changes requested.

The-Huginn commented:
Please keep this one on hold. I will update the readme manual for openshift deployment and test it.

xstefank commented on README.md:
```diff
@@ -101,6 +120,9 @@ wildfly:
    title-check:
      pattern: "\\[WFLY-\\d+\\]\\s+.*|WFLY-\\d+\\s+.*"
      message: "Wrong content of the title!"
+ emails:
```
we should probably call it only email, so not to expect multiple emails

The-Huginn replied:
We have for example `regexes` under description and we allow only single regex. Isn't it the same?

xstefank commented on README.md:
```diff
@@ -43,6 +43,10 @@ QUARKUS_GITHUB_APP_WEBHOOK_PROXY_URL=<your Smee.io channel URL>
+
+QUARKUS_MAILER_FROM=<email address>
+QUARKUS_MAILER_USERNAME=<email address>
+QUARKUS_MAILER_PASSWORD=<email password>
```
shouldn't we mention that this is optional if you don't send emails?

The-Huginn replied:
Adding a note

The-Huginn commented:
Additional changes were introduced:

* Resolve #74 
* Check for rules missing id - this one was removed previously.
* Change to catching `JsonProcessingException` instead of `JsonMappingException`
* Added and updated tests in `StartupTest`

xstefank commented on src/main/java/io/xstefank/wildfly/bot/ConfigFileChangeProcessor.java:
```diff
@@ -46,14 +50,20 @@ void onFileChanged(@PullRequest.Opened @PullRequest.Edited @PullRequest.Synchron
+                            githubCommitProcessor.commitStatusSuccess(pullRequest, CHECK_NAME, "Valid");
+                            Log.debug("Configuration File check successful");
+                        } else {
+                            githubCommitProcessor.commitStatusError(pullRequest, CHECK_NAME, "\u274C Rule is missing an id or multiple rules have the same id.");
```
We removed icons in different PR

xstefank commented on src/main/java/io/xstefank/wildfly/bot/LifecycleProcessor.java:
```diff
@@ -46,8 +57,26 @@ void onStart(@Observes StartupEvent event) {
+                            LOG.errorf("The configuration file from the repository %s was not parsed successfully due to following problems: %s", repository.getFullName(), problems);
+
+                            if (username.isPresent() && emailAddresses != null && !emailAddresses.isEmpty()) {
+                                LOG.infof("Sending email to the following people [%s].", String.join(", ", emailAddresses));
```
```suggestion
                                LOG.infof("Sending email to the following emails [%s].", String.join(", ", emailAddresses));
```

xstefank commented on src/main/java/io/xstefank/wildfly/bot/LifecycleProcessor.java:
```diff
@@ -46,8 +57,26 @@ void onStart(@Observes StartupEvent event) {
+                                mailer.send(
+                                        new Mail()
+                                                .setSubject("Unsuccessful installation of Wildfly Bot Application")
+                                                .setText("Hello,\n\nThe configuration file " + RuntimeConstants.CONFIG_FILE_NAME + " has some invalid rules in the following github repository: " + repository.getHttpTransportUrl() + " . The following problems were detected. " + problems)
```
add something in line of "this is generated message, do not respond"

xstefank commented on src/main/java/io/xstefank/wildfly/bot/LifecycleProcessor.java:
```diff
@@ -46,8 +57,26 @@ void onStart(@Observes StartupEvent event) {
+                                                .setTo(emailAddresses)
+                                );
+                            } else {
+                                LOG.warn("No emails setup to receive warnings or no email address setup to send emails from.");
```
since this is optional, we don't need Warn, do a debug

xstefank commented on src/main/resources/application.properties:
```diff
@@ -37,3 +37,12 @@ LW6V8BbcP5MRhd2JQSRh16nWwSQJ2BdpUZFwayEEQ6UcrMfqvA0=\
+
+# Sending emails
+quarkus.mailer.auth-methods=DIGEST-MD5 CRAM-SHA256 CRAM-SHA1 CRAM-MD5 PLAIN LOGIN
+quarkus.mailer.host=smtp.gmail.com
```
should this be hardcoded in properties?

The-Huginn replied:
It's there so the user does not need to write those properties, if they decide to use this feature

xstefank commented on src/test/java/io/xstefank/wildfly/bot/StartupEventTest.java:
```diff
@@ -51,43 +64,141 @@ record -> record.getLevel().intValue() >= Level.INFO.intValue());
+
+        private final String configFile;
+
+        private static final String[] configRules = new String[] {
```
does this needs to be rewritten?

The-Huginn reviewed (commented):
Additionally to PR review fixes, I have updated README

xstefank commented on src/main/java/io/xstefank/wildfly/bot/ConfigFileChangeProcessor.java:
```diff
@@ -46,14 +50,20 @@ void onFileChanged(@PullRequest.Opened @PullRequest.Edited @PullRequest.Synchron
+                            githubCommitProcessor.commitStatusSuccess(pullRequest, CHECK_NAME, "Valid");
+                            Log.debug("Configuration File check successful");
+                        } else {
+                            githubCommitProcessor.commitStatusError(pullRequest, CHECK_NAME, "Rule is missing an id or multiple rules have the same id.");
```
why not pass problems into the status too?

The-Huginn replied:
It would be unreadable in the status info I would assume. That's why it sends only recommendation. We might extract which problems are present form the return `List` and log only relevant problems.

xstefank replied:
even a string this long won't fit into the status window and that is expected. If you hover your mouse over the text, it shows it in full.

xstefank commented on src/main/java/io/xstefank/wildfly/bot/LifecycleProcessor.java:
```diff
@@ -46,8 +57,27 @@ void onStart(@Observes StartupEvent event) {
+                                mailer.send(
+                                        new Mail()
+                                                .setSubject("Unsuccessful installation of Wildfly Bot Application")
+                                                .setText("Hello,\n\nThe configuration file " + RuntimeConstants.CONFIG_FILE_NAME + " has some invalid rules in the following github repository: " + repository.getHttpTransportUrl() +
```
btw, why don't you try String.format with multiline string for better readability?

xstefank commented on src/test/java/io/xstefank/wildfly/bot/StartupEventTest.java:
```diff
@@ -51,43 +64,141 @@ record -> record.getLevel().intValue() >= Level.INFO.intValue());
+    void init() {
+        mailbox.clear();
+    }
+
```
delete one empty line

xstefank commented on src/test/java/io/xstefank/wildfly/bot/StartupEventTest.java:
```diff
@@ -51,43 +64,141 @@ record -> record.getLevel().intValue() >= Level.INFO.intValue());
+                    List<Mail> sent = mailbox.getMailsSentTo("foo@bar.baz");
+                    Assertions.assertEquals(sent.size(), 1);
+                    Assertions.assertEquals(sent.get(0).getSubject(), "Unsuccessful installation of Wildfly Bot Application");
+                    Assertions.assertEquals(sent.get(0).getText(), "Hello,\n\nThe configuration file " + RuntimeConstants.CONFIG_FILE_NAME + " has some invalid rules in the following github repository: " + repository.getHttpTransportUrl() + " . The following problems were detected. [Rule [id=null title=Test body=null titleBody=null directories=[] notify=[0979986727, 7125767235]] is missing an id] ...
```
same here

xstefank commented on src/main/java/io/xstefank/wildfly/bot/ConfigFileChangeProcessor.java:
```diff
@@ -46,14 +50,20 @@ void onFileChanged(@PullRequest.Opened @PullRequest.Edited @PullRequest.Synchron
+                            Log.debug("Configuration File check successful");
+                        } else {
+                            githubCommitProcessor.commitStatusError(pullRequest, CHECK_NAME, String.join(",", problems));
+                            Log.warnf("Configuration File check unsuccessful. %s", String.join(",", problems));
```
nothing big, but if you do the same operation in two lines after each other it's better to extract the result. I don't expect problems will be a large list but still.

xstefank commented on src/main/java/io/xstefank/wildfly/bot/LifecycleProcessor.java:
```diff
@@ -46,8 +57,29 @@ void onStart(@Observes StartupEvent event) {
+                                        new Mail()
+                                                .setSubject("Unsuccessful installation of Wildfly Bot Application")
+                                                .setText(String.format("""
+                                                        Hello,\n
```
This will add two new lines, are you aware of that?

The-Huginn replied:
Yes, I didn't want to use `\n` at all, but I had to because of the `maven-checkstyle-plugin` didn't accept empty lines in string

xstefank commented on src/test/java/io/xstefank/wildfly/bot/StartupEventTest.java:
```diff
@@ -51,43 +64,146 @@ record -> record.getLevel().intValue() >= Level.INFO.intValue());
+            GHAuthenticatedAppInstallation mockGHAuthenticatedAppInstallation = mock(GHAuthenticatedAppInstallation.class);
+            PagedSearchIterable<GHRepository> mockGHRepositories = mock(PagedSearchIterable.class);
+            PagedIterator<GHRepository> mockIterator = mock(PagedIterator.class);
+            mocks.configFile("wildfly-bot.yml").fromString(configFile);
```
we could use the constant here too

xstefank reviewed (approved):
LGTM but fix the small issue mentioned
